### PR TITLE
first preview of treebuilder and task queue

### PR DESCRIPTION
**Pull request by Kiyoshika**
This PR introduces the first rough draft of a "tree builder" interface and a task queue to queue up changes to the DOM and commit them at once.

The general structure at a glance is as follows:
* A `TreeBuilder` interface (trait) is created which provides operations that can be performed on the DOM
* A `DocumentTaskQueue` implements this trait where each method creates and appends a `DocumentTask` containing the necessary information to make changes later
* An `execute()` method is provided in `DocumentTaskQueue` which will iterate all pending changes and commit them to the actual DOM. The queue is then emptied and tasks can be inserted again for further modification.
* Any failures that occur are appended into a vector (I added a new error type to @emwalker's types) and returned from `execute()` (or empty vector if none occurred)
* When creating nodes with the interface, temporary IDs are created and returned which will be the real IDs those nodes will have if the changes are committed, but they are not added to the node arena unless `execute()` is called. This is managed by an internal counter and added for convenience if the parser decides to use it to manage the `open_elements`

This `DocumentTaskQueue` is what can be passed/used by a parser which is essentially the abstracted interface to manipulate the DOM without the parser needing to handle the implementation details - those are handled by `DocumentTaskQueue` which interacts with `DocumentHandle` during `execute()`

For a somewhat involved example, see the [unit test](https://github.com/gosub-browser/gosub-engine/compare/main...feat/tree-builder-interface#diff-b6445b24b00c4c34cf6ad8fc67a26db7ddfcf284bfb1b0e0c73c2da8874e9f66R917)

Right now this implementation coexists with the current code and does not actually touch the parser or any of the core functionality - it only serves as an alternative to manipulating the DOM. This PR is just intended introduce the concept into the codebase and we can slowly fine tune it and merge it into the parser once it's ready. There may definitely need to be more work before we can actually use it in the parser

### CODE REVIEW

**Review by jaytaph — APPROVED** [review]

File: src/html5/parser/document.rs [hunk 1]
```diff
@@ -60,6 +62,153 @@ impl DocumentFragment {
...
+    fn insert_attribute(&mut self, key: &str, value: &str, element_id: NodeId) {
+        let attribute = DocumentTask::InsertAttribute {
+            key: key.to_owned(),
+            value: value.to_owned(),
+            element_id,
+        };
+        self.tasks.push(attribute);
+    }
+
+    fn execute(&mut self) -> Vec<String> {
```
Comment (emwalker) [inline src/html5/parser/document.rs:164]: I think "flush" is the term I've been seeing at the level of the queue of tasks.

**Review by emwalker — APPROVED** [review]
LGTM with thoughts / suggestions.

File: src/html5/parser/document.rs [hunk 2]
```diff
@@ -60,6 +62,153 @@ impl DocumentFragment {
...
+    },
+    CreateComment {
+        content: String,
+        parent_id: NodeId,
+    },
+    InsertAttribute {
+        key: String,
+        value: String,
+        element_id: NodeId,
+    },
```
[the remark this answers is not in the document]
Comment (Kiyoshika) [reply src/html5/parser/document.rs:90]: I think as of right now in our codebase Task is unambiguous but if we for some reason need a different structure that is "task" related then at that point we could think of a rename

File: src/html5/parser/document.rs (hunk 1 above)
Comment (Kiyoshika) [reply src/html5/parser/document.rs:164]: I was swapping back and forth between `flush` and `execute`, I don't really have a strong opinion so I'm open to renaming

File: src/html5/parser/document.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (emwalker) [reply src/html5/parser/document.rs:90]: Sounds good.

**Pull request merged**